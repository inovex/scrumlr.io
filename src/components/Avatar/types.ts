export const AVATAR_SKIN_COLORS = ["Tanned", "Yellow", "Pale", "Light", "Brown", "DarkBrown", "Black"] as const;
export type AvatarSkinColor = (typeof AVATAR_SKIN_COLORS)[number];

export const AVATAR_TOP_TYPES = [
  "Eyepatch",
  "Hat",
  "Hijab",
  "LongHairBigHair",
  "LongHairBob",
  "LongHairBun",
  "LongHairCurly",
  "LongHairCurvy",
  "LongHairDreads",
  "LongHairFrida",
  "LongHairFro",
  "LongHairFroBand",
  "LongHairMiaWallace",
  "LongHairNotTooLong",
  "LongHairShavedSides",
  "LongHairStraight",
  "LongHairStraight2",
  "LongHairStraightStrand",
  "NoHair",
  "ShortHairDreads01",
  "ShortHairDreads02",
  "ShortHairFrizzle",
  "ShortHairShaggyMullet",
  "ShortHairShortCurly",
  "ShortHairShortFlat",
  "ShortHairShortRound",
  "ShortHairShortWaved",
  "ShortHairSides",
  "ShortHairTheCaesar",
  "ShortHairTheCaesarSidePart",
  "Turban",
  "WinterHat1",
  "WinterHat2",
  "WinterHat3",
  "WinterHat4",
] as const;
export type AvatarTopType = (typeof AVATAR_TOP_TYPES)[number];

export const AVATAR_CLOTHE_COLORS = [
  "Black",
  "Blue01",
  "Blue02",
  "Blue03",
  "Gray01",
  "Gray02",
  "Heather",
  "PastelBlue",
  "PastelGreen",
  "PastelOrange",
  "PastelRed",
  "PastelYellow",
  "Pink",
  "Red",
  "White",
] as const;
export type AvatarClotheColor = (typeof AVATAR_CLOTHE_COLORS)[number];

export const AVATAR_GRAPHIC_TYPES = ["Bat", "Cumbia", "Deer", "Diamond", "Hola", "Pizza", "Resist", "Selena", "Bear", "SkullOutline", "Skull"] as const;
export type AvatarGraphicType = (typeof AVATAR_GRAPHIC_TYPES)[number];

export const AVATAR_CLOTHE_TYPES = ["BlazerShirt", "BlazerSweater", "CollarSweater", "GraphicShirt", "Hoodie", "Overall", "ShirtCrewNeck", "ShirtScoopNeck", "ShirtVNeck"] as const;
export type AvatarClotheType = (typeof AVATAR_CLOTHE_TYPES)[number];

export const AVATAR_HAIR_COLORS = ["Auburn", "Black", "Blonde", "BlondeGolden", "Brown", "BrownDark", "PastelPink", "Blue", "Platinum", "Red", "SilverGray"] as const;
export type AvatarHairColor = (typeof AVATAR_HAIR_COLORS)[number];

export const AVATAR_FACIAL_HAIR_COLORS = ["Auburn", "Black", "Blonde", "BlondeGolden", "Brown", "BrownDark", "Platinum", "Red"] as const;
export type AvatarFacialHairColor = (typeof AVATAR_FACIAL_HAIR_COLORS)[number];

export const AVATAR_FACIAL_HAIR_TYPES = ["Blank", "BeardMedium", "BeardLight", "BeardMajestic", "MoustacheFancy", "MoustacheMagnum"] as const;
export type AvatarFacialHairType = (typeof AVATAR_FACIAL_HAIR_TYPES)[number];

export const AVATAR_ACCESSORIES_TYPES = ["Blank", "Kurt", "Prescription01", "Prescription02", "Round", "Sunglasses", "Wayfarers"] as const;
export type AvatarAccessoriesType = (typeof AVATAR_ACCESSORIES_TYPES)[number];

export const AVATAR_EYE_TYPES = ["Close", "Cry", "Default", "Dizzy", "EyeRoll", "Happy", "Hearts", "Side", "Squint", "Surprised", "Wink", "WinkWacky"] as const;
export type AvatarEyeType = (typeof AVATAR_EYE_TYPES)[number];

export const AVATAR_EYEBROW_TYPES = [
  "Angry",
  "AngryNatural",
  "Default",
  "DefaultNatural",
  "FlatNatural",
  "FrownNatural",
  "RaisedExcited",
  "RaisedExcitedNatural",
  "SadConcerned",
  "SadConcernedNatural",
  "UnibrowNatural",
  "UpDown",
  "UpDownNatural",
] as const;
export type AvatarEyebrowType = (typeof AVATAR_EYEBROW_TYPES)[number];

export const AVATAR_MOUTH_TYPES = ["Concerned", "Default", "Disbelief", "Eating", "Grimace", "Sad", "ScreamOpen", "Serious", "Smile", "Tongue", "Twinkle", "Vomit"] as const;
export type AvatarMouthType = (typeof AVATAR_MOUTH_TYPES)[number];
