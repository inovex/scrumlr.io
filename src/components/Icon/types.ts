export type IconNames =
  | 'phase1-create'
  | 'phase1-communicate'
  | 'phase1-combine'
  | 'phase2-vote'
  | 'phase3-discuss'
  | 'phase3-define'
  | 'circle-arrow-left'
  | 'circle-arrow-right'
  | 'circle-add'
  | 'circle'
  | 'more'
  | 'plus'
  | 'circle-selection'
  | 'circle-selection-grey'
  | 'logout'
  | 'download'
  | 'edit'
  | 'donate'
  | 'pencil'
  | 'focus'
  | 'close'
  | 'close20'
  | 'close-circle'
  | 'feedback'
  | 'key-up'
  | 'key-down'
  | 'key-esc'
  | 'settings'
  | 'board'
  | 'trash'
  | 'chevron-left'
  | 'chevron-right'
  | 'stack-top'
  | 'stack-mid'
  | 'stack-hover'
  | 'arrow-left'
  | 'arrow-right';
