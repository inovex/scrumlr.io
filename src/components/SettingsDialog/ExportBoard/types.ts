import {Color} from "constants/colors";

export interface BoardType {
  id: string;
  name: string;
  showAuthors: boolean;
  showVoting: string;
  accessPolicy: string;
}

export interface ColumnType {
  id: string;
  name: string;
  color: Color;
  visible: boolean;
}

export interface NoteType {
  id: string;
  author: string;
  text: string;
  position: {
    column: string;
    stack: string | null;
    rank: number;
  };
}

export interface ParticipantType {
  role: string;
  user: {
    id: string;
    name: string;
  };
}

export interface VotingType {
  votes: {
    votes: number;
    votesPerNote: [];
  };
}

export interface BoardDataType {
  board: BoardType;
  columns: ColumnType[];
  notes: NoteType[];
  participants: ParticipantType[];
  votings: VotingType[];
}
