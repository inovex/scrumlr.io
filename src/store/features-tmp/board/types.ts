import {Participant} from "../../features/participants/types";
import {Request} from "../../features/requests/types";
import {Column} from "../../features/columns/types";
import {Note} from "../../features/notes/types";
import {Reaction} from "../../features/reactions/types";
import {Vote} from "../../features/votes/types";
import {Voting} from "../../features/votings/types";

export enum AccessPolicy {
  "PUBLIC" = 0,
  "BY_PASSPHRASE" = 1,
  "BY_INVITE" = 2,
}

export interface Board {
  id: string;

  name?: string;
  accessPolicy: keyof typeof AccessPolicy;
  showAuthors: boolean;
  showNotesOfOtherUsers: boolean;
  showNoteReactions: boolean;
  allowStacking: boolean;
  isLocked: boolean;
  timerStart?: Date;
  timerEnd?: Date;

  sharedNote?: string;
  showVoting?: string;
}

export type BoardActionType = {
  board: Board;
  participants: Participant[];
  requests: Request[];
  columns: Column[];
  notes: Note[];
  reactions: Reaction[];
  votes: Vote[];
  votings: Voting[];
};

export type EditBoardRequest = Partial<Omit<Board, "id">> & {passphrase?: string};

export type BoardStatus = "unknown" | "pending" | "ready" | "rejected" | "accepted" | "passphrase_required" | "incorrect_passphrase" | "too_many_join_requests" | "banned";

export interface BoardState {
  status: BoardStatus;
  data?: Board;
}
