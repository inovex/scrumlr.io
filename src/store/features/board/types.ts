import {Participant} from "../participants";
import {Request} from "../requests";
import {Column} from "../columns";
import {Note} from "../notes";
import {Reaction} from "../reactions";
import {Vote} from "../votes";
import {Voting} from "../votings";

export enum AccessPolicy {
  "PUBLIC" = 0,
  "BY_PASSPHRASE" = 1,
  "BY_INVITE" = 2,
}

export interface Board {
  id: string;

  name?: string;
  accessPolicy: keyof typeof AccessPolicy;
  showAuthors: boolean;
  showNotesOfOtherUsers: boolean;
  showNoteReactions: boolean;
  allowStacking: boolean;
  isLocked: boolean;
  timerStart?: Date;
  timerEnd?: Date;

  sharedNote?: string;
  showVoting?: string;
}

export type BoardActionType = {
  board: Board;
  participants: Participant[];
  requests: Request[];
  columns: Column[];
  notes: Note[];
  reactions: Reaction[];
  votes: Vote[];
  votings: Voting[];
};

export type BoardWithServerTimeOffset = {board: Board; serverTimeOffset: number};

export type EditBoardRequest = Partial<Omit<Board, "id">> & {passphrase?: string};

export type BoardStatus = "unknown" | "pending" | "ready" | "rejected" | "accepted" | "passphrase_required" | "incorrect_passphrase" | "too_many_join_requests" | "banned";

export interface BoardState {
  status: BoardStatus;
  data?: Board;
}
