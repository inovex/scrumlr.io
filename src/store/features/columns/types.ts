import {Color} from "constants/colors";

export interface Column {
  id: string;
  name: string;
  description: string;
  color: Color;
  visible: boolean;
  index: number;
}

export type ColumnWithoutId = Omit<Column, "id">;

export type ColumnsState = Column[];
