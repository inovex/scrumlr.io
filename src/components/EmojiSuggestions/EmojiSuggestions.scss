@import "constants/style";

.emoji-suggestions {
  padding: $padding--small;
  border-radius: $rounded--default;
  background-color: $color-lighter-gray;
  list-style: none;
  margin: 0;

  color: $color-black;

  &__element {
    cursor: pointer;
    overflow: hidden;
    white-space: nowrap;
    text-overflow: ellipsis;
    padding: $padding--small;
    border-radius: $rounded--small;
    line-height: $line-height--medium;
  }

  &__element:hover,
  .emoji-suggestions__element--focus {
    background-color: rgba(var(--accent-color-rgb), 0.2);
  }

  &__emoji {
    margin-right: $margin--small;
    user-select: none;
    font-size: $text--md;
  }
}

[theme="dark"] {
  .emoji-suggestions {
    color: $color-white;
    background-color: $color-dark-mode-note;

    &__element:hover,
    .emoji-suggestions__element--focus {
      background-color: #485064;
    }
  }
}
