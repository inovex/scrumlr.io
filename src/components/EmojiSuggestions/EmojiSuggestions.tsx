import classNames from "classnames";
import {EmojiData, emojiWithSkinTone} from "utils/hooks/useEmojiAutocomplete";
import "./EmojiSuggestions.scss";
import {useEffect, useRef} from "react";
import {useAppSelector} from "store";

type EmojiSuggestionsProps = {
  suggestions: EmojiData[];
  keyboardFocusedIndex: number;
  acceptSuggestion: (insertedEmoji: string) => void;
};

export const EmojiSuggestions = ({suggestions, keyboardFocusedIndex, acceptSuggestion}: EmojiSuggestionsProps) => {
  // the refs of the elements to scroll into view
  const suggestionsRef = useRef<HTMLLIElement[]>([]);

  useEffect(() => {
    // whenever the suggestions change, remove the old refs
    suggestionsRef.current = suggestionsRef.current.slice(0, suggestions.length);

    // scroll the currently selected element into view
    suggestionsRef.current[keyboardFocusedIndex]?.scrollIntoView({block: "nearest"});
  }, [keyboardFocusedIndex, suggestions]);

  const skinToneComponent = useAppSelector((state) => state.skinTone.component);

  if (suggestions.length === 0) return null;

  return (
    <ul className="emoji-suggestions">
      {suggestions.map(([slug, emoji, supportsSkintones], i) => {
        const actualEmoji = supportsSkintones ? emojiWithSkinTone(emoji, skinToneComponent) : emoji;
        return (
          <li
            className={classNames("emoji-suggestions__element", {"emoji-suggestions__element--focus": i === keyboardFocusedIndex})}
            onMouseDown={(e) => {
              // using onMouseDown to prevent the input from losing focus
              e.preventDefault();
              acceptSuggestion(actualEmoji);
            }}
            key={slug}
            ref={(el) => {
              suggestionsRef.current[i] = el!;
            }}
          >
            <span className="emoji-suggestions__emoji">{actualEmoji}</span>:{slug}:
          </li>
        );
      })}
    </ul>
  );
};
