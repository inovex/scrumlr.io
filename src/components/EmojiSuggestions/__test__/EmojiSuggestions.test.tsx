import {Provider} from "react-redux";
import {ApplicationState} from "types";
import getTestStore from "utils/test/getTestStore";
import {ComponentProps} from "react";
import {render} from "@testing-library/react";
import {EmojiData} from "utils/hooks/useEmojiAutocomplete";
import {EmojiSuggestions} from "../EmojiSuggestions";

type EmojiSuggestionsProps = ComponentProps<typeof EmojiSuggestions>;

const mockEmojis: EmojiData[] = [
  ["face_with_tears_of_joy", "😂", false, ["face_with_tears_of_joy", "face", "cry", "tears", "weep", "happy", "happytears", "haha"]],
  ["sparkling_heart", "💖", false, ["sparkling_heart", "love", "like", "affection", "valentines"]],
  ["thumbs_up", "👍", true, ["thumbs_up", "thumbsup", "yes", "awesome", "good", "agree", "accept", "cool", "hand", "like"]],
  ["woman_pilot", "👩‍✈️", true, ["woman_pilot", "aviator", "plane", "woman", "human"]],
];

function createSuggestions(props: Partial<EmojiSuggestionsProps> = {}, state?: Partial<ApplicationState>) {
  const suggestionProps: EmojiSuggestionsProps = {
    acceptSuggestion: jest.fn(),
    keyboardFocusedIndex: 0,
    suggestions: mockEmojis,
    ...props,
  };

  return (
    <Provider store={getTestStore(state)}>
      <EmojiSuggestions {...suggestionProps} />
    </Provider>
  );
}

describe("EmojiSuggestions", () => {
  beforeEach(() => {
    Element.prototype.scrollIntoView = jest.fn();
  });

  it("should match the snapshot", () => {
    const {container} = render(createSuggestions());
    expect(container.firstChild).toMatchSnapshot();
  });

  it("should accept suggestion on click", () => {
    const acceptSuggestion = jest.fn();
    const {container} = render(
      createSuggestions({
        acceptSuggestion,
      })
    );

    const element = container.getElementsByClassName("emoji-suggestions__element");
    element.item(0)?.dispatchEvent(new MouseEvent("mousedown", {bubbles: true}));

    expect(acceptSuggestion).toHaveBeenCalledWith("😂");
  });

  describe("should use skin tones where supported", () => {
    it("should use skin tone from store", () => {
      const {container} = render(createSuggestions({}, {skinTone: {name: "medium", component: "🏽"}}));
      expect(container.firstChild).toMatchSnapshot();
    });

    it("should pass skin tone to acceptSuggestions", () => {
      const acceptSuggestion = jest.fn();
      const {container} = render(
        createSuggestions(
          {
            acceptSuggestion,
          },
          {skinTone: {name: "medium", component: "🏽"}}
        )
      );

      const element = container.getElementsByClassName("emoji-suggestions__element");

      element.item(2)?.dispatchEvent(new MouseEvent("mousedown", {bubbles: true}));
      expect(acceptSuggestion).toHaveBeenCalledWith("👍🏽");

      element.item(3)?.dispatchEvent(new MouseEvent("mousedown", {bubbles: true}));
      expect(acceptSuggestion).toHaveBeenCalledWith("👩🏽‍✈️");
    });
  });
});
