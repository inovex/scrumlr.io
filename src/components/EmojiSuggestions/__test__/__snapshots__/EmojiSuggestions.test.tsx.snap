// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EmojiSuggestions should match the snapshot 1`] = `
<ul
  class="emoji-suggestions"
>
  <li
    class="emoji-suggestions__element emoji-suggestions__element--focus"
  >
    <span
      class="emoji-suggestions__emoji"
    >
      😂
    </span>
    :
    face_with_tears_of_joy
    :
  </li>
  <li
    class="emoji-suggestions__element"
  >
    <span
      class="emoji-suggestions__emoji"
    >
      💖
    </span>
    :
    sparkling_heart
    :
  </li>
  <li
    class="emoji-suggestions__element"
  >
    <span
      class="emoji-suggestions__emoji"
    >
      👍
    </span>
    :
    thumbs_up
    :
  </li>
  <li
    class="emoji-suggestions__element"
  >
    <span
      class="emoji-suggestions__emoji"
    >
      👩‍✈️
    </span>
    :
    woman_pilot
    :
  </li>
</ul>
`;

exports[`EmojiSuggestions should use skin tones where supported should use skin tone from store 1`] = `
<ul
  class="emoji-suggestions"
>
  <li
    class="emoji-suggestions__element emoji-suggestions__element--focus"
  >
    <span
      class="emoji-suggestions__emoji"
    >
      😂
    </span>
    :
    face_with_tears_of_joy
    :
  </li>
  <li
    class="emoji-suggestions__element"
  >
    <span
      class="emoji-suggestions__emoji"
    >
      💖
    </span>
    :
    sparkling_heart
    :
  </li>
  <li
    class="emoji-suggestions__element"
  >
    <span
      class="emoji-suggestions__emoji"
    >
      👍🏽
    </span>
    :
    thumbs_up
    :
  </li>
  <li
    class="emoji-suggestions__element"
  >
    <span
      class="emoji-suggestions__emoji"
    >
      👩🏽‍✈️
    </span>
    :
    woman_pilot
    :
  </li>
</ul>
`;
